### PR TITLE
HoH applications support

### PR DESCRIPTION
**Pull request by golanshabi**

### CODE REVIEW

**Comment by nirrozenbaum** [comment]
soon we're going to release a new version v0.2.0.
apps is not part of this version and therefore I suggest to wait with this PR until version is released and tested.
(should be release in the next few days)

File: deploy/operator.yaml.template [hunk 1]
```diff
@@ -99,6 +99,33 @@ rules:
...
+  - update
+- apiGroups:
+  - "apps.open-cluster-management.io"
+  resources:
+  - channels
+  verbs:
+  - get
+  - list
+  - watch
+  - update
```
Comment (nirrozenbaum) [inline deploy/operator.yaml.template:119]: this section is not needed.
just put `- channels` in line 106

File: go.mod [hunk 2]
```diff
@@ -91,4 +91,13 @@ require (
 	sigs.k8s.io/yaml v1.2.0 // indirect
 )
 
+require (
+	github.com/go-test/deep v1.0.7 // indirect
+	github.com/open-cluster-management/hub-of-hubs-data-types v0.1.1-0.20211025070346-60d0cacb5152 // indirect
+	github.com/open-cluster-management/multicloud-operators-channel v1.0.1-0.20201120143200-e505a259de45
+	github.com/open-cluster-management/multicloud-operators-subscription v1.2.2-2-20201130-59f96
```
Comment (nirrozenbaum) [inline go.mod:98]: these are old versions and are not the ones from ACM v2.4

File: go.mod [hunk 3]
```diff
@@ -91,4 +91,13 @@ require (
...
+	golang.org/x/tools v0.1.5 // indirect
+	sigs.k8s.io/application v0.8.3
```
Comment (nirrozenbaum) [inline go.mod:100]: which application version is used in ACM v2.4?

File: pkg/controller/controllers.go [hunk 4]
```diff
@@ -8,19 +8,25 @@ import (
 
 	"github.com/jackc/pgx/v4/pgxpool"
 	policiesv1 "github.com/open-cluster-management/governance-policy-propagator/api/v1"
+	chanv1 "github.com/open-cluster-management/multicloud-operators-channel/pkg/apis/apps/v1"
 	appsv1 "github.com/open-cluster-management/multicloud-operators-placementrule/pkg/apis/apps/v1"
+	subv1 "github.com/open-cluster-management/multicloud-operators-subscription/pkg/apis/apps/v1"
 	configv1 "github.com/stolostron/hub-of-hubs-data-types/apis/config/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	applicationv1 "sigs.k8s.io/application/api/v1beta1"
```
Comment (nirrozenbaum) [inline pkg/controller/controllers.go:16]: let's try to use readable names as alias.
channlesv1, subscriptionsv1, appsv1beta1
the current alias appsv1 should be placementrulesv1

File: pkg/controller/controllers.go [hunk 5]
```diff
@@ -8,19 +8,25 @@ import (
...
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/scheme"
 )
 
 // AddToScheme adds all the resources to be processed to the Scheme.
-func AddToScheme(s *runtime.Scheme) error {
-	schemeBuilders := []*scheme.Builder{policiesv1.SchemeBuilder, appsv1.SchemeBuilder, configv1.SchemeBuilder}
+func AddToScheme(sch *runtime.Scheme) error {
```
Comment (nirrozenbaum) [inline pkg/controller/controllers.go:22]: what is variable name sch?
this is not readable.

according to best practices, we either give a readable name or a one letter name as it was before this change

File: pkg/controller/application_spec_sync.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
+package controller
+
+import (
+	"fmt"
+
+	"github.com/jackc/pgx/v4/pgxpool"
+	"k8s.io/apimachinery/pkg/api/equality"
+	appsv1 "sigs.k8s.io/application/api/v1beta1"
```
Comment (nirrozenbaum) [inline pkg/controller/application_spec_sync.go:8]: this alias may be confusing.
should be `appsv1beta1` (v1beta1 is the version.. beta makes it clear it's still not production level package)

File: pkg/controller/application_spec_sync.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func areApplicationsEqual(instance1, instance2 client.Object) bool {
```
Comment (nirrozenbaum) [inline pkg/controller/application_spec_sync.go:13]: for the sake of code organization and consistency, I would use the same order as in other spec sync files.
that is - first the add controller, then clean status func, and then are equal func.
see for example here - https://github.com/stolostron/hub-of-hubs-spec-sync/blob/main/pkg/controller/placementbinding_spec_sync.go

File: pkg/controller/application_spec_sync.go [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+		return fmt.Errorf("failed to add Application Controller to the manager: %w", err)
+	}
+
+	return nil
+}
+
+func cleanApplicationStatus(instance client.Object) {
+	application, ok := instance.(*appsv1.Application)
+	if !ok {
+		panic("wrong instance passed to cleanConfigStatus: not appsv1.Application")
```
Comment (nirrozenbaum) [inline pkg/controller/application_spec_sync.go:46]: looks like a copy paste typo.. cleanConfigStatus is only in the config spec sync.
all others should log their function name.. e.g. here it's `cleanApplicationStatus`

File: pkg/controller/channel_spec_sync.go [hunk 9]
```diff
@@ -0,0 +1,51 @@
+package controller
+
+import (
+	"fmt"
+
+	"github.com/jackc/pgx/v4/pgxpool"
+	chanv1 "github.com/open-cluster-management/multicloud-operators-channel/pkg/apis/apps/v1"
```
Comment (nirrozenbaum) [inline pkg/controller/channel_spec_sync.go:7]: channelsv1

File: pkg/controller/channel_spec_sync.go [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+	"k8s.io/apimachinery/pkg/api/equality"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func cleanChannelStatus(instance client.Object) {
+	channel, ok := instance.(*chanv1.Channel)
+	if !ok {
+		panic("wrong instance passed to cleanConfigStatus: not chanv1.Channel")
```
Comment (nirrozenbaum) [inline pkg/controller/channel_spec_sync.go:16]: typo in logging

File: pkg/controller/channel_spec_sync.go [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+import (
+	"fmt"
+
+	"github.com/jackc/pgx/v4/pgxpool"
+	chanv1 "github.com/open-cluster-management/multicloud-operators-channel/pkg/apis/apps/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
```
Comment (nirrozenbaum) [inline pkg/controller/channel_spec_sync.go:12]: order of functions.. see comment in application spec sync

File: pkg/controller/channel_spec_sync.go [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+		Complete(&genericSpecToDBReconciler{
+			client:                 mgr.GetClient(),
+			databaseConnectionPool: databaseConnectionPool,
+			log:                    ctrl.Log.WithName("channel-spec-syncer"),
+			tableName:              "channels",
+			finalizerName:          "hub-of-hubs.open-cluster-management.io/channel-cleanup",
+			createInstance:         func() client.Object { return &chanv1.Channel{} },
+			cleanStatus:            cleanChannelStatus,
+			areEqual:               areChannelsEqual,
+			shouldProcess:          namespaceFilterFunc,
```
Comment (nirrozenbaum) [inline pkg/controller/channel_spec_sync.go:44]: this is not the way to filter events.
for doing filtering you can use controller predicate and define what are the conditions that the controller will get the events.
I think you're already familiar with the controller predicate mechanism we have in leaf hub status sync (same as you did for local policies)

File: pkg/controller/channel_spec_sync.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+		})
+	if err != nil {
+		return fmt.Errorf("failed to add Channel Controller to the manager: %w", err)
```
Comment (nirrozenbaum) [inline pkg/controller/channel_spec_sync.go:47]: no need for capital letters in log

File: pkg/controller/application_spec_sync.go [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+			databaseConnectionPool: databaseConnectionPool,
+			log:                    ctrl.Log.WithName("application-spec-syncer"),
+			tableName:              "applications",
+			finalizerName:          "hub-of-hubs.open-cluster-management.io/application-cleanup",
+			createInstance:         func() client.Object { return &appsv1.Application{} },
+			cleanStatus:            cleanApplicationStatus,
+			areEqual:               areApplicationsEqual,
+		})
+	if err != nil {
+		return fmt.Errorf("failed to add Application Controller to the manager: %w", err)
```
Comment (nirrozenbaum) [inline pkg/controller/application_spec_sync.go:37]: no need for capital letters in log

File: pkg/controller/subscription_spec_sync.go [hunk 15]
```diff
@@ -0,0 +1,59 @@
+package controller
+
+import (
+	"fmt"
+
+	"github.com/jackc/pgx/v4/pgxpool"
+	subv1 "github.com/open-cluster-management/multicloud-operators-subscription/pkg/apis/apps/v1"
```
Comment (nirrozenbaum) [inline pkg/controller/subscription_spec_sync.go:7]: subscriptionsv1

File: pkg/controller/subscription_spec_sync.go [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+	"k8s.io/apimachinery/pkg/api/equality"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
```
Comment (nirrozenbaum) [inline pkg/controller/subscription_spec_sync.go:12]: functions order, `cleanConfigStatus` typo, remove the `shouldProcess` bool and replace with predicate.
if you want to use namespace name in several places, put it in some consts/helpers file and not in one of the places that uses it.
I can ask in this case - why did you put the const in subscription file and not in channel file?
we need to be consistent with our coding standards.

File: pkg/controller/subscription_spec_sync.go [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+			log:                    ctrl.Log.WithName("subscriptions-spec-syncer"),
+			tableName:              "subscriptions",
+			finalizerName:          "hub-of-hubs.open-cluster-management.io/subscription-cleanup",
+			createInstance:         func() client.Object { return &subv1.Subscription{} },
+			cleanStatus:            cleanSubscriptionStatus,
+			areEqual:               areSubscriptionsEqual,
+			shouldProcess:          namespaceFilterFunc,
+		})
+	if err != nil {
+		return fmt.Errorf("failed to add Subscription Controller to the manager: %w", err)
```
Comment (nirrozenbaum) [inline pkg/controller/subscription_spec_sync.go:30]: no need for capital letters in log

File: pkg/controller/generic_spec_to_db_reconciler.go [hunk 18]
```diff
@@ -27,6 +27,7 @@ type genericSpecToDBReconciler struct {
 	createInstance         func() client.Object
 	cleanStatus            func(client.Object)
 	areEqual               func(client.Object, client.Object) bool
+	shouldProcess          func(client.Object) bool
```
Comment (nirrozenbaum) [inline pkg/controller/generic_spec_to_db_reconciler.go:30]: this should be removed.
use controller predicate instead.

**Review by nirrozenbaum — COMMENTED** [review]
@golanshabi I've added some comments.
in addition `go.sum` here is huge and should be recreated.

File: go.mod (hunk 3 above)
Comment (golanshabi) [reply go.mod:100]: checked again its the correct version

**Review by nirrozenbaum — APPROVED** [review]

**Pull request merged**